File: src/chapter7_linear/best_time_to_buy_and_sell_stock/BestTimeToBuyAndSellStock.java
package chapter7_linear.best_time_to_buy_and_sell_stock;

public class BestTimeToBuyAndSellStock {

    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int minPrice = prices[0];

        for (int price : prices) {
            minPrice = Math.min(minPrice, price);
            maxProfit = Math.max(price - minPrice, maxProfit);
        }

        return maxProfit;
    }
}

File: src/chapter7_linear/best_time_to_buy_and_sell_stock/BestTimeToBuyAndSellStock.kt
package chapter7_linear.best_time_to_buy_and_sell_stock

fun maxProfit(prices: IntArray): Int {
    var maxProfit = 0
    var minPrice = prices[0]

    for (price in prices) {
        minPrice = minPrice.coerceAtMost(price)
        maxProfit = maxProfit.coerceAtLeast(price - minPrice)
    }

    return maxProfit
}